### PR TITLE
✨ Add Qwen3.5 LLM

### PR DESCRIPTION
**Pull request by LBerth**
Add Qwen3.5 LLM: 
- Qwen3.5 code is nearly identical to [Rashka's github](https://github.com/rasbt/LLMs-from-scratch/tree/main/ch05/16_qwen3.5), modulo some cleaning and fix bad practices (assert -> raise, use dataclass for settings, put all related functions inside Qwen class)
- Put Qwen's tokenizer in tokenizers module + store the tokenizer json file inside repo
- Add a unit test
- Uses KV cache
- Passes Ruff and mypy
- Fix a typo in GPT2

⚠️ I couldn't make the fast implementation (using flash linear attention) work, but I kept the imports in the beginning, in case we can make it work in the future.

### CODE REVIEW

File: mfai/tokenizers/__init__.py [hunk 1]
```diff
@@ -201,3 +203,93 @@ def eot_token(self) -> int:
...
+        add_generation_prompt: bool = False,
+        add_thinking: bool = False,
+    ) -> None:
+        self.apply_chat_template = apply_chat_template
+        self.add_generation_prompt = add_generation_prompt
+        self.add_thinking = add_thinking
+
+        tokenizer_file = Path(__file__).parent / "qwen3.5/tokenizer.json"
+        self._tok = tokenizers.Tokenizer.from_file(str(tokenizer_file))
+        self._special_to_id = {}
```
Comment (kalharko) [inline mfai/tokenizers/__init__.py]:
```suggestion
        self._tok: list[int] = tokenizers.Tokenizer.from_file(str(tokenizer_file))
        self._special_to_id: dict[str, int] = {}
```

File: mfai/pytorch/models/llms/qwen3_5.py [hunk 2]
```diff
@@ -0,0 +1,1246 @@
...
+    )
+
+    use_fast_conv1d = True
+except ImportError:
+    warnings.warn(
+        "The fast conv1d is not available because the required library is not "
+        "installed. Falling back to torch implementation. "
+        "To install follow https://github.com/Dao-AILab/causal-conv1d."
+    )
+use_fast_conv1d = False
```
Comment (kalharko) [inline mfai/pytorch/models/llms/qwen3_5.py]: `use_fast_conv1d` should be set to False before the try-except code block ?

if true, also applies to `use_flash_att`

File: mfai/pytorch/models/llms/qwen3_5.py [hunk 3]
```diff
@@ -0,0 +1,1246 @@
...
+    see https://github.com/state-spaces/mamba/issues/66
+    """
+    # NOTE: attention mask is a 2D boolean tensor
+    if (
+        attention_mask is not None
+        and attention_mask.shape[1] > 1
+        and attention_mask.shape[0] > 1
+    ):
+        dtype = hidden_states.dtype
+        hidden_states = (hidden_states * attention_mask[:, :, None]).to(dtype)
```
Comment (kalharko) [inline mfai/pytorch/models/llms/qwen3_5.py:122]: Should't we convert the `attention_mask` to the `hidden_states.dtype` before multiplying it ?
We risk precision loss on the `hidden_states` if the mask's dtype is less precise than the hidden state's one ?

Comment (LBerth) [reply mfai/pytorch/models/llms/qwen3_5.py:122]: as the attention mask is a boolean tensor, the values from hidden states will just be multiplied by 0 or 1, so I don't think there's an issue here.

**Review by colon3ltocard — APPROVED** [review]

**Review by kalharko — APPROVED** [review]
